### ISSUE TITLE
Look in `Data\BashTags` for additional Bash Tags applied to plugins

### ISSUE THREAD

**Issue by Ortham** [loot/loot#1408]
As brought up in loot/skyrimse#1359, a plugin's description field is limited to 512 bytes and Wrye Bash allows mod authors to list Bash Tags that can't fit in the description field in `Data\BashTags`. It would be good if LOOT could check there for additional tags when loading a plugin's current Bash Tags.

**Comment by Infernio** [comment]
Keep in mind that Wrye Bash lets those win over LOOT. This is because the feature was originally intended for guide authors, so they could supply users with tags that are specifically added for the guide and don't make sense for a normal user, so shouldn't be submitted to the LOOT masterlist.

At least, that was the intention. The actual usages I've seen in practice look like this:
- [Lexy's LOTD guide](https://lexyslotd.com/) uses them as intended, for adding tags specific to Lexy's guide to plugins.
- [This mod](https://www.nexusmods.com/oblivion/mods/49849) uses them for adding newer tags to old plugins. Most of these could be submitted to the LOOT masterlist, some have already been added in https://github.com/loot/oblivion/issues/112.
- A new use case that wasn't anticipated at all is that the description might become too small, as mentioned above. This leaves mod authors with only two choices: submit all their tags to the masterlist, or use a BashTags file.
- Anecdotally, I use them for adding tags to plugins before testing and submitting them to the LOOT masterlist.

So my question is: *should* they keep winning over LOOT, or should we move them so they win over the description but lose against LOOT?

Relevant links:
- WB's documentation on these files: https://wrye-bash.github.io/docs/Wrye%20Bash%20Advanced%20Readme.html#patch-tags
- The issue where they were requested: https://github.com/wrye-bash/wrye-bash/issues/463

**Comment by Infernio** [comment]
One more thing, the online docs are only up to date for 307. One important change to the specification for these files since then is that
```
Note: BashTags files must be in UTF-8 encoding (ASCII is fine, since it is backwards-compatible).
```

**Comment by Ortham** [comment]
> So my question is: _should_ they keep winning over LOOT, or should we move them so they win over the description but lose against LOOT?

I think that an application should generally prefer data sourced from its own store over external stores, so Wrye Bash should prefer what's in BashTags files over LOOT's masterlist/userlist.

The obvious problem with that thinking is that it means LOOT should prefer its masterlist/userlist over BashTags files, which would be wrong. That's because Bash Tags are a Wrye Bash feature, so it's what WB does that actually matters. 

However, following that thought, LOOT doesn't display Bash Tags to show the user what WB is going to do: it shows what Bash Tags are in the plugin's description, and what LOOT suggests should be added or removed. If BashTags files have greater precedence than LOOT's metadata, there's no benefit to LOOT supporting them in any way, because they aren't an input into LOOT's suggestions.

I flip-flopped a bit on whether or not it's even useful for LOOT to display Bash Tag suggestions in its UI, since if WB can find LOOT it automatically applies them, but WB may not be able to find LOOT, so it's probably still useful to keep them displayed.

**Comment by Infernio** [comment]
Maybe just show some sort of indicator then when a BashTags file is found for a plugin? Just to let users know that the tags they set in LOOT there might not be the ones they actually end up with in WB?

**Comment by Ortham** [comment]
> Maybe just show some sort of indicator then when a BashTags file is found for a plugin? Just to let users know that the tags they set in LOOT there might not be the ones they actually end up with in WB?

The trouble with that is that LOOT doesn't know if WB can find its masterlists, so doesn't know if *any* of its suggestions will actually end up in WB.

**Comment by MacSplody** [comment]
The latest update for [Unofficial Skyrim Special Edition Patch](https://www.nexusmods.com/skyrimspecialedition/mods/266), the tags are now stored in `Data\Bash_Tags`.

```
# Generated by Wrye Bash 308.202012142034
Actors.ACBS, Actors.AIData, Actors.AIPackages, Actors.CombatStyle, Actors.DeathItem, Actors.Spells, Actors.Stats, C.Acoustic, C.Climate, C.Encounter, C.ImageSpace, C.Light, C.Location, C.LockList, C.Music, C.Name, C.Owner, C.Regions, C.Water, Delev, EffectStats, EnchantmentStats, Factions, Graphics, Invent.Add, Invent.Change, Invent.Remove, Keywords, NPC.Class, NPC.Race, Names, ObjectBounds, Outfits.Add, Outfits.Remove, Relations.Add, Relations.Change, Relev, Sound, SpellStats, Stats, Text
```

https://github.com/loot/skyrimse/pull/1850/commits/392d41421623c169fb8e1eae0277c23284c7ffc2

I've gone with not including the tags listed in `Data\Bash_Tags`, since the user can't change them through LOOT anyway.
If we wanted to display tags listed in `Data\Bash_Tags` then it would make sense to disable removal of those tags through LOOT.

**Comment by sibir-ine** [comment]
@Infernio This may be excessive, but maybe it'd be worth adding support for another file that can serve as an extension of the Bash Tags list in the description for cases like USSEP's?

Something like:
```
Data
|- BashTagsBase
|  |- Mod Name.txt
```

It would function the same as the files in `Data\BashTags\`, but it's priority would instead be between the plugin's description & `taglist.yaml`. That way, mods like USSEP can use this new file to store default Bash Tags instead of in a file with the second-to-highest priority, people can continue to use the file at `Data\BashTags\` appropriately, & LOOT could read this new file without worrying about the implications of it having higher priority than it's own master/userlists.

**Comment by Infernio** [comment]
The problem is that each of those files has an impact on startup. We need to query the filesystem on whether or not it exists, once per plugin the user has installed. Once WB's refreshes are event-based (https://github.com/wrye-bash/wrye-bash/issues/265) this would be more reasonable to add.

**Comment by sibir-ine** [comment]
Ah, I understand. Then, until such a feature is more reasonable (or a better alternative is proposed), I agree with Ortham that they should continue to override LOOT's suggestions & that they shouldn't be displayed in LOOT's interface. However, I do agree with Infernio that it could be useful to display if one exists for a plugin & warn users to add/remove tags through WB instead of through LOOT (not sure how many people add/remove tags through their userlists instead of WB in the first place so 🤷 ).

**Comment by MacSplody** [comment]
> (not sure how many people add/remove tags through their userlists instead of WB in the first place so 🤷 ).

I do it through LOOT because it doesn't affect Wrye Bash's automatic processing of tags.

**Comment by Infernio** [comment]
WB *combines* tags from these sources:
 1. plugin description
 2. LOOT taglist (included with WB, used only if no LOOT masterlist could be found)
 3. LOOT masterlist
 4. LOOT userlist
 5. `Data\BashTags`

That means your example is handled like this:
 1. nothing in the plugin description -> {}
 2. LOOT masterlist present, skip taglist -> {}
 3. LOOT masterlist adds C.Water and Delev -> {C.Water, Delev}
 4. No LOOT userlist present -> {C.Water, Delev}
 5. BashTags file removes C.Water -> {Delev}

Result: {Delev}

The only thing that overrides tags completely is when you set them manually in WB. All other sources get combined/merged together like this.

**Comment by Ortham** [comment]
OK, thanks for the clarification.

In that case, when loading plugin (meta)data, LOOT could check if it's got any Bash Tag suggestions for a plugin, and if so check if the plugin has a file in `Data\BashTags`. If it does, it could then parse the file to get Bash Tags from it, and then add a warning message if any of those are removals of Bash Tags that are suggested for addition by LOOT (or vice versa), naming the affected Bash Tags in the message.

**Comment by Ortham** [comment]
I've implemented the above in the `bash-tags-files` branch. The message that gets displayed is of the form:

> This plugin has a BashTags file that will override the suggestions made by LOOT for the following Bash Tags: C.Water, Factions.

I'm open to suggestions for a better way to word that (or a different message entirely).

**Comment by sibir-ine** [comment]
Maybe the message could be a note instead of a warning for users who intentionally override some suggestions for their specific mod list. That &/or a filter to hide the messages for the same reason.

**Comment by Ortham** [comment]
> Maybe the message could be a note instead of a warning for users who intentionally override some suggestions for their specific mod list. That &/or a filter to hide the messages for the same reason.

I've made the message a note instead of a warning.

**Comment by Ortham** [comment]
Merged as of c5582ec89db375bdeeb95893908697206a961a84.
